src/systems/elevator/state_machine/StatesImpl.hh

```diff
@@ -0,0 +1,245 @@
+  {
+  }
+
+  /// \brief Destructor
+  public: virtual ~DoorState()
+  {
+  }
+
+  /// \brief Sends the opening or closing command to the door joint of the
+  /// target floor and configures the monitor that checks the joint state
+  /// \param[in] _fsm State machine to which the state belongs
+  public: template <typename Event, typename FSM>
+  void on_enter(const Event &, FSM &_fsm)
+  {
+    const auto &data = _fsm.Data();
+    int32_t floorTarget = data->system->state;
+    ignmsg << "The elevator is " << this->Report(data) << std::endl;
+
+    double jointTarget = this->JointTarget(data, floorTarget);
+    data->SendCmd(data->system->doorJointCmdPubs[floorTarget], jointTarget);
+    this->triggerEvent = [&_fsm] { _fsm.process_event(E()); };
+    data->system->SetDoorMonitor(
+        floorTarget, jointTarget, this->posEps, this->velEps,
+        std::bind(&DoorState::OnJointTargetReached, this));
+  }
+
+  /// \brief Gets the message that's being reported when entering the state
+  /// \param[in] _data State machine data
+  /// \return String with the message the state has to report
+  private: virtual std::string Report(
+      const std::unique_ptr<ElevatorStateMachinePrivate> &_data) const = 0;
+
+  /// \brief Gets the joint position for opening or closing a door
+  /// \param[in] _data State machine data
+  /// \param[in] _floorTarget Target floor level
+  /// \return Joint position
+  private: virtual double JointTarget(
+      const std::unique_ptr<ElevatorStateMachinePrivate> &_data,
+      int32_t _floorTarget) const = 0;
+
+  /// \brief Method that gets called when a door joint reaches its target
+  private: void OnJointTargetReached() { this->triggerEvent(); }
+
+  /// \brief Positiion tolerance
+  private: double posEps;
+
+  /// \brief Velocity tolerance
+  private: double velEps;
+
+  /// \brief Triggers the exit event
+  private: std::function<void()> triggerEvent;
+};
+
+/// \brief State at which the elevator is opening a door.
+struct ElevatorStateMachineDef::OpenDoorState : DoorState<events::DoorOpen>
+{
+  /// \brief This state defers EnqueueNewTargetEvent events
+  public: using deferred_events = type_tuple<events::EnqueueNewTarget>;
+
+  /// \brief Constructor
+  public: OpenDoorState() : DoorState(5e-2)
+  {
+  }
+
+  // Documentation inherited
+  private: virtual std::string Report(
+      const std::unique_ptr<ElevatorStateMachinePrivate> &_data) const override
+  {
+    return "opening door " + std::to_string(_data->system->state);
+  }
+
+  // Documentation inherited
+  private: virtual double JointTarget(
+      const std::unique_ptr<ElevatorStateMachinePrivate> &_data,
+      int32_t _floorTarget) const override
+  {
+    return _data->system->doorTargets[_floorTarget];
+  }
+};
+
+/// \brief State at which the elevator is closing a door.
+struct ElevatorStateMachineDef::CloseDoorState : DoorState<events::DoorClosed>
+{
+  /// \brief Constructor
+  public: CloseDoorState() : DoorState(2e-2)
```

Comment:
Is there a reason to override the tolerance here? Maybe this should be the default value?

Reply:
I think I have the open state at 5cm because it takes some "wasted" time to reach steady-state, and the close state at 2cm because, otherwise, I was seeing the cabin move before the door was "fully" closed. I can probably live with both at 2cm.